File: ImageAlignment/image_alignment_simple_example.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video/tracking.hpp>

#include <cmath>
#include <filesystem>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>

namespace fs = std::filesystem;

struct Options {
    fs::path template_path = "images/image1.jpg";
    fs::path moving_path = "images/image2.jpg";
    fs::path output = "output/image2-aligned.jpg";
    std::string motion = "euclidean";
    int iterations = 5000;
    double epsilon = 1e-7;
    bool display = true;
    bool validate = false;
};

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int i = 1; i < argc; ++i) {
        const std::string argument = argv[i];
        if (argument == "--template" && i + 1 < argc) {
            options.template_path = argv[++i];
        } else if (argument == "--moving" && i + 1 < argc) {
            options.moving_path = argv[++i];
        } else if (argument == "--output" && i + 1 < argc) {
            options.output = argv[++i];
        } else if (argument == "--motion" && i + 1 < argc) {
            options.motion = argv[++i];
        } else if (argument == "--iterations" && i + 1 < argc) {
            options.iterations = std::stoi(argv[++i]);
        } else if (argument == "--epsilon" && i + 1 < argc) {
            options.epsilon = std::stod(argv[++i]);
        } else if (argument == "--no-display") {
            options.display = false;
        } else if (argument == "--validate") {
            options.validate = true;
        } else if (argument == "--help") {
            std::cout
                << "Usage: image_alignment_simple [--template FILE] "
                   "[--moving FILE] [--motion translation|euclidean|affine|homography] "
                   "[--output FILE] [--no-display] [--validate]\n";
            std::exit(0);
        } else {
            throw std::invalid_argument("Unknown or incomplete argument: " + argument);
        }
    }
    return options;
}

double meanAbsoluteError(const cv::Mat& first, const cv::Mat& second) {
    cv::Mat difference;
    cv::absdiff(first, second, difference);
    return cv::mean(difference)[0] + cv::mean(difference)[1] +
           cv::mean(difference)[2];
}

int main(int argc, char** argv) {
    try {
        const Options options = parseOptions(argc, argv);
        const std::map<std::string, int> motion_models = {
            {"translation", cv::MOTION_TRANSLATION},
            {"euclidean", cv::MOTION_EUCLIDEAN},
            {"affine", cv::MOTION_AFFINE},
            {"homography", cv::MOTION_HOMOGRAPHY},
        };
        const auto motion_iterator = motion_models.find(options.motion);
        if (motion_iterator == motion_models.end()) {
            throw std::invalid_argument("Unknown motion model: " + options.motion);
        }
        const int motion_model = motion_iterator->second;

        const cv::Mat template_image = cv::imread(options.template_path.string());
        cv::Mat moving_image = cv::imread(options.moving_path.string());
        if (template_image.empty() || moving_image.empty()) {
            throw std::runtime_error("Could not read one or both input images");
        }
        if (template_image.size() != moving_image.size()) {
            cv::resize(moving_image, moving_image, template_image.size(), 0.0, 0.0,
                       cv::INTER_AREA);
        }
        cv::Mat template_gray;
        cv::Mat moving_gray;
        cv::cvtColor(template_image, template_gray, cv::COLOR_BGR2GRAY);
        cv::cvtColor(moving_image, moving_gray, cv::COLOR_BGR2GRAY);
        template_gray.convertTo(template_gray, CV_32F);
        moving_gray.convertTo(moving_gray, CV_32F);

        cv::Mat warp = motion_model == cv::MOTION_HOMOGRAPHY
                           ? cv::Mat::eye(3, 3, CV_32F)
                           : cv::Mat::eye(2, 3, CV_32F);
        const cv::TermCriteria criteria(
            cv::TermCriteria::COUNT | cv::TermCriteria::EPS, options.iterations,
            options.epsilon);
        const double correlation =
            cv::findTransformECC(template_gray, moving_gray, warp, motion_model,
                                 criteria);

        cv::Mat aligned;
        const int flags = cv::INTER_LINEAR | cv::WARP_INVERSE_MAP;
        if (motion_model == cv::MOTION_HOMOGRAPHY) {
            cv::warpPerspective(moving_image, aligned, warp, template_image.size(),
                                flags);
        } else {
            cv::warpAffine(moving_image, aligned, warp, template_image.size(),
                           flags);
        }
        const double before = meanAbsoluteError(template_image, moving_image);
        const double after = meanAbsoluteError(template_image, aligned);
        if (!options.output.parent_path().empty()) {
            fs::create_directories(options.output.parent_path());
        }
        if (!cv::imwrite(options.output.string(), aligned)) {
            throw std::runtime_error("Could not write aligned image");
        }

        std::cout << "OpenCV: " << CV_VERSION << "\n"
                  << "Motion model: " << options.motion << "\n"
                  << "ECC correlation: " << correlation << "\n"
                  << "Warp matrix:\n" << warp << "\n"
                  << "Mean absolute error: " << before << " -> " << after << "\n"
                  << "Saved: " << options.output << "\n";

        if (options.validate) {
            if (!cv::checkRange(warp) || !std::isfinite(correlation) ||
                correlation <= 0.0 || after >= before) {
                throw std::runtime_error("ECC pair-alignment validation failed");
            }
            std::cout << "Validation: PASS\n";
        }
        if (options.display) {
            cv::imshow("Template", template_image);
            cv::imshow("Moving", moving_image);
            cv::imshow("Aligned", aligned);
            cv::waitKey(0);
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << "\n";
        return 1;
    }
}

File: ImageAlignment/image_alignment_simple_example.py
#!/usr/bin/env python3
"""Align one image to another with OpenCV's ECC optimizer."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2
import numpy as np

from ecc_utils import MOTION_MODELS, align_image, mean_absolute_error


SCRIPT_DIR = Path(__file__).resolve().parent


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--template", type=Path, default=SCRIPT_DIR / "images" / "image1.jpg")
    parser.add_argument("--moving", type=Path, default=SCRIPT_DIR / "images" / "image2.jpg")
    parser.add_argument(
        "--motion", choices=sorted(MOTION_MODELS), default="euclidean"
    )
    parser.add_argument("--iterations", type=int, default=5000)
    parser.add_argument("--epsilon", type=float, default=1e-7)
    parser.add_argument(
        "--output", type=Path, default=SCRIPT_DIR / "output" / "image2-aligned.jpg"
    )
    parser.add_argument("--no-display", action="store_true")
    parser.add_argument("--validate", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    template = cv2.imread(str(args.template))
    moving = cv2.imread(str(args.moving))
    if template is None:
        raise FileNotFoundError(f"Could not read template: {args.template}")
    if moving is None:
        raise FileNotFoundError(f"Could not read moving image: {args.moving}")

    correlation, warp, aligned = align_image(
        template,
        moving,
        MOTION_MODELS[args.motion],
        args.iterations,
        args.epsilon,
    )
    moving_for_comparison = (
        moving
        if moving.shape[:2] == template.shape[:2]
        else cv2.resize(
            moving,
            (template.shape[1], template.shape[0]),
            interpolation=cv2.INTER_AREA,
        )
    )
    before = mean_absolute_error(template, moving_for_comparison)
    after = mean_absolute_error(template, aligned)
    args.output.parent.mkdir(parents=True, exist_ok=True)
    if not cv2.imwrite(str(args.output), aligned):
        raise RuntimeError(f"Could not write output: {args.output}")

    print(f"OpenCV: {cv2.__version__}")
    print(f"Motion model: {args.motion}")
    print(f"ECC correlation: {correlation:.8f}")
    print(f"Warp matrix:\n{warp}")
    print(f"Mean absolute error: {before:.3f} -> {after:.3f}")
    print(f"Saved: {args.output}")

    if args.validate:
        if (
            not np.isfinite(warp).all()
            or not np.isfinite(correlation)
            or correlation <= 0.0
            or after >= before
        ):
            raise RuntimeError("ECC pair-alignment validation failed")
        print("Validation: PASS")

    if not args.no_display:
        cv2.imshow("Template", template)
        cv2.imshow("Moving", moving_for_comparison)
        cv2.imshow("Aligned", aligned)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
